azurelinuxagent/ga/persist_firewall_rules.py

```diff
@@ -101,67 +98,60 @@ def _is_firewall_service_running():
+            logger.info("Firewalld service not running/unavailable, trying to set up {0}".format(self._network_setup_service_name))
+            if systemd.is_systemd():
+                self._setup_network_setup_service()
+            else:
+                event.info(WALAEventOperation.PersistFirewallRules, "Did not detect systemd, will not set {0}", self._network_setup_service_name)
             return
 
-        logger.info(
-            "Firewalld service not running/unavailable, trying to set up {0}".format(self._network_setup_service_name))
+        event.info(WALAEventOperation.PersistFirewallRules,"Firewalld.service is running, setting up permanent rules on the VM")
+        # In case of a failure, this would throw. In such a case, we don't need to try to setup our custom service
+        # because on system reboot, all iptables rules are reset by firewalld.service, so it would be a no-op.
+        # setup permanent firewalld rules
+        firewall_manager = FirewallCmd(self._dst_ip)
 
-        self._setup_network_setup_service()
+        firewall_manager.remove_legacy_rule()
 
-    def __verify_firewall_rules_enabled(self):
-        # Check if firewall-rules have already been enabled
-        # This function would also return False if the dest-ip is changed. So no need to check separately for that
         try:
-            AddFirewallRules.check_firewalld_rule_applied(self._dst_ip, self._uid)
-        except Exception as error:
-            logger.verbose(
-                "Check if Firewall rules already applied using firewalld.service failed: {0}".format(ustr(error)))
-            return False
-
-        return True
-
-    def __remove_firewalld_rules(self):
+            if firewall_manager.check():
+                event.info(WALAEventOperation.PersistFirewallRules, "The permanent firewall rules for Azure Fabric are already setup:\n{0}", firewall_manager.get_state())
+            else:
+                firewall_manager.setup()
+                event.info(WALAEventOperation.PersistFirewallRules, "Created permanent firewall rules for Azure Fabric:\n{0}", firewall_manager.get_state())
+        except FirewallStateError as e:
+            event.warn(
+                WALAEventOperation.PersistFirewallRules,
+                "The permanent firewall rules for Azure Fabric are not setup correctly ({0}), will reset them. Current state:\n{1}", ustr(e), firewall_manager.get_state())
+            firewall_manager.remove()
+            firewall_manager.setup()
+            event.info(WALAEventOperation.PersistFirewallRules, "Reset permanent firewall rules for Azure Fabric:\n{0}", firewall_manager.get_state())
+
+        # Remove custom service if exists to avoid problems with firewalld
         try:
-            AddFirewallRules.remove_firewalld_rules(self._dst_ip, self._uid)
+            fileutil.rm_files(*[self.get_service_file_path(), os.path.join(conf.get_lib_dir(), self.BINARY_FILE_NAME)])
```

Comment:
Adding here somewhat confusing to me. since we setup before firewallmanger and then I had to figure there is return to believe that we don't come to here when firewallmanager not running. May be move above as part of custom firewall setup

Reply:
Not sure I understand the concern. I think the intention of the original code was to remove the file when initially firewalld was not running and the Agent created the custom service, then on a subsequent Agent start, firewalld is running and we need to delete the file.

Reply:
I understand the intention, my comment on organizing the code. I would prefer to place it before we setup the firewall rules with firewalld. That wat it looks cleaner

Reply:
I think I'm missing something here. Do you mean moving the delete before line 132? that makes no difference as far as code readability/organization, I think.